### DISCUSSION TITLE
Proposed future Arbiter architecture

### DISCUSSION THREAD

**Discussion by ts0yu** [primitivefinance/arbiter#295] [Ideas]
Colin said I should post this here :)

![image-3.png](https://github.com/primitivefinance/arbiter/assets/120932697/2b5bc595-a808-4f6b-841f-2d717b832746)

![image-4.png](https://github.com/primitivefinance/arbiter/assets/120932697/3d0056cd-f53b-40c2-bfab-813eb70a2a96)

**Comment by Autoparallel** [comment]
@kinrezC we should think about database stuff in here.

**Comment by 0xJepsen** [comment]
@ts0yu I think this is a really good approach. I think Q3 this year we will be focussing on really cleaning up the code base. We have finished a simple sim for uniswap which we are just polishing up before we do a write up for.

**Comment by ts0yu** [comment]
Thanks. Regarding the database stuff @Autoparallel , I was thinking we should have some "Inspector" trait you could attach to the engine. The trait would have a log function, and this means users could adapt and change how their simulations return results.

**Reply by Autoparallel** [reply]
Yeah I think this is good as a trait. I think toggling on and off logging of quantities associated to agents would be very slick.

**Comment by Autoparallel** [comment]
I was just thinking... any time we have to do "pure math" (i.e., calling `solstat` or `solmate` functions) in the EVM, we could avoid doing `transact_commit()`. In fact, we could build an interface for these math functions that has it's own EVM that implements `transact()` only and is just used for this sole purpose.

**Comment by ts0yu** [comment]
The other thing to think about is how to make this "env agnostic"... So how could we make this available for anvil, revm, mainnet, and any other user defined state capture mechanism.

**Reply by Autoparallel** [reply]
My thought on this is building out very minimal `ethers::middleware` or `client` for revm. Doing this has other benefits as far as using bindings as well. This would be huge, but I'm not really sure how to do it. I spent a few hours in middleware hell and called it quits. I'd like to revisit it though.

This point is huge though. I agree strongly.
